pickhardtpayments/Payment.py

```diff
@@ -0,0 +1,207 @@
+        """Set time when payment was finished, either by being aborted or by successful settlement
+
+        Sets end_time time in seconds from epoch. Should be called when the Payment failed or when Payment
+        settled successfully.
+
+        :param timestamp: time in seconds from epoch
+        :type timestamp: float
+        """
+        self._end_time = timestamp
+
+    @property
+    def attempts(self):
+        """Returns all onions that were built and are associated with this Payment.
+
+        :return: A list of Attempts of this payment.
+        :rtype: list[Attempt]
+        """
+        return self._attempts
+
+    def add_attempts(self, attempts: list[Attempt]):
+        """Adds Attempts (onions) that have been made to settle the Payment to the Payment object.
+
+        :param attempts: a list of attempts that belong to this Payment
+        :type: list[Attempt]
+        """
+        self._attempts.extend(attempts)
+
+    @property
+    def fee(self):
+        """Returns the fees that accrued for this payment. It's the sum of the routing fees of all settled onions.
+
+        :return: fee in sats for successful attempts of Payment
+        :rtype: float
+        """
+        fee = 0
+        for attempt in self.settled_attempts:
+            fee += attempt.routing_fee
+        return fee
+
+    @property
+    def ppm(self):
+        """Returns the fees that accrued for this payment. It's the sum of the routing fees of all settled onions.
+
+        :return: fee in ppm for successful delivery and settlement of payment
+        :rtype: float
+        """
+        return self.fee * 1000 / self.total_amount
+
+    @property
+    def inflight_attempts(self):
+        """Returns all onions that can successfully be routed.
+
+        :return: A list of successful Attempts of this Payment, which could be settled.
+        :rtype: list[Attempt]
+        """
+        inflight_attempts = []
+        try:
+            for attempt in self._attempts:
+                if attempt.status == Attempt.SettlementStatus.INFLIGHT:
+                    inflight_attempts.append(attempt)
+            return inflight_attempts
+        except ValueError:
+            logging.warning("ValueError in Payment.inflight_attempts")
+        return []
+
+    @property
+    def failed_attempts(self):
+        """Returns all onions that failed, where no delivery of the (partial) amount was possible.
+
+        :return: A list of failed Attempts of this Payment, Attempts that could be not settled.
+        :rtype: list[Attempt]
+        """
+        failed_attempts = []
+        try:
+            for attempt in self._attempts:
+                if attempt.status == Attempt.SettlementStatus.FAILED:
+                    failed_attempts.append(attempt)
+            return failed_attempts
+        except ValueError:
+            logging.warning("ValueError in Payment.failed_attempts")
+            return []
+
+    @property
+    def settled_attempts(self):
```

Comment:
why not using a `filter_attempt` and provide `Attempt.SettlementStatus.SETTELD` as an argument?

Reply:
yes - as discussed; changed it